### ISSUE TITLE
System.ValueTuple reference missing from F# compiler default references when using .NET 4.7.1

### ISSUE THREAD

**Issue by dsyme** [Microsoft/visualfsharp#3896]
I'm not sure of the cause but I'm seeing three test failures on my local machine. All three are related to ``System.ValueTuple.dll``.  Note this machine has .NET 4.7.1 installed. 

The individual steps are:

    cd tests\fsharp\typecheck\sigs
    ..\..\release\net40\bin\fsc.exe -r:System.Core.dll --nowarn:20 --define:COMPILED --target:exe -o:pos27.exe pos27.fs

    cd tests\fsharp\core\quotes
    ..\..\release\net40\bin\fsc.exe -r:System.Core.dll --nowarn:20 --define:COMPILED -o:test.exe -r cslib.dll -g test.fsx

Both appear to be corrected by adding ``-r:System.ValueTuple.dll`` to the compilation line.  I think it's just a problem with the test framework

Screenshot of errors below

![image](https://user-images.githubusercontent.com/7204669/32456448-8a8b127c-c31d-11e7-886b-f3384af29cd8.png)

**Comment by Pilchie** [comment]
Tagging @terrajobst

**Comment by dsyme** [comment]
@Pilchie @terrajobst  FWIW I do think this is just a test suite error.  I know there's a lot of stuff flying around at the moment about System.ValueTuple and ,NET 4.7.1 but I don't yet think this need worry you

**Comment by terrajobst** [comment]
That's a [known bug in .NET Framework 4.7.1](https://github.com/dotnet/standard/issues/567). We'll push a fix later. /cc @AlexGhiondea

**Comment by dsyme** [comment]
> That's a known bug in .NET Framework 4.7.1. We'll push a fix later. /cc @AlexGhiondea

Yep, I know about that issue but I don't think this is a symptom of it - I think in this case it's just a test suite error of some kind :)

**Comment by terrajobst** [comment]
Sounds good. Considering there is no error message I'm sure you're correct :-)

**Comment by KevinRansom** [comment]
@dsyme the only thing wrong with the test issue theory is that those tests have run for ages without this failure.  So ... either the compiler has changed how it magically references things or something else has changed ... I did see this at the weekend, and will be looking it at.  But I'm not really sure why it happens yet.  Or even how to reliably repro it.

**Comment by dsyme** [comment]
@KevinRansom If my understanding is correct, System.ValueTuple is in mscorlib.dll on 4.7.1.  I suspect that when running fsc.exe on 4.7.1 without ``--noframework`` (i.e. using default references) we generate a default reference to the runtime copy of mscorlib.dll [here](https://github.com/Microsoft/visualfsharp/blob/master/src/fsharp/CompileOps.fs#L1780).  Presumably we've already generated a different reference to mscorlib earlier as the simple reference text "mscorlib.dll" which is I believe resolved to the default latest known framework reference assembly rather than an implementation assembly.

In short our technique to generate a default reference for System.ValueTuple.dll doesn't work when running the compiler on 4.7.1

We should fix it somehow.  However I don't think it's that urgent
* Using default references is fairly rare  (it happens, including in our test suite, direct uses of fsc.exe, and direct uses of FSharp.Compiler.Service's Compile method)
* Using System.ValueTuple is rare 
* Using both is really pretty rare, and there is a simple workaround

**Comment by KevinRansom** [comment]
Yes, that is why I was reluctant to blame the test.  Anyway .. it is something I was planning on looking at.

**Comment by dsyme** [comment]
- @KevinRansom Thanks. I believe that always referencing the System.ValueTuple.dll that we ship alongside fsc.exe (if it is present) would be adequate.
